File: backspace-string-compare/backspace-string-compare.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var backspaceCompare = function(s, t) {
    function makeString(str) {
        let ans = [];
        for (let i=0;i<str.length;i++) {
            if (str[i] == '#' && i > 0) {
                ans.pop();
            }
            else if(str[i] != '#') {
                ans.push(str[i]);
            }
        }
        //console.log(`${str}: ${ans}`);
        return ans.join('');
    }
    return makeString(s) == makeString(t);
};

File: backspace-string-compare/backspace-string-compare.java
class Solution {
    public boolean backspaceCompare(String s, String t) {
        Stack<Character> sStk = new Stack();
        Stack<Character> tStk = new Stack();
        for (char c : s.toCharArray()) {
            if (c == '#') {
                if (sStk.size() > 0) {
                    sStk.pop();
                }
            }
            else {
                sStk.push(c);
            }
        }
        for (char c : t.toCharArray()) {
            if (c == '#') {
                if (tStk.size() > 0) {
                    tStk.pop();
                }
            }
            else {
                tStk.push(c);
            }
        }
        if (sStk.size() != tStk.size()) {
            return false;
        }
        while (sStk.size() > 0) {
            if (sStk.pop() != tStk.pop()) {
                return false;
            } 
        }
        return true;
    }
}